<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>TrIdent - Transduction Identification • TrIdent</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="TrIdent - Transduction Identification">
<meta name="description" content="The `TrIdent` R package automates the analysis of transductomics data by detecting, classifying, and characterizing read coverage patterns associated with potential transduction events. Transductomics is a DNA sequencing-based method for the detection and characterization of transduction events in pure cultures and complex communities. Transductomics relies on mapping sequencing reads from a viral-like particle (VLP)-fraction of a sample to contigs assembled from the metagenome (whole-community) of the same sample. Reads from bacterial DNA carried by VLPs will map back to the bacterial contigs of origin creating read coverage patterns indicative of ongoing transduction.">
<meta property="og:description" content="The `TrIdent` R package automates the analysis of transductomics data by detecting, classifying, and characterizing read coverage patterns associated with potential transduction events. Transductomics is a DNA sequencing-based method for the detection and characterization of transduction events in pure cultures and complex communities. Transductomics relies on mapping sequencing reads from a viral-like particle (VLP)-fraction of a sample to contigs assembled from the metagenome (whole-community) of the same sample. Reads from bacterial DNA carried by VLPs will map back to the bacterial contigs of origin creating read coverage patterns indicative of ongoing transduction.">
<meta property="og:image" content="https://jlmaier12.github.io/TrIdent/logo.png">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">TrIdent</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/TrIdent-vignette.html">TrIdent</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/jlmaier12/TrIdent/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header">
<img src="logo.png" class="logo" alt=""><h1 id="trident">TrIdent<a class="anchor" aria-label="anchor" href="#trident"></a>
</h1>
</div>
<p><strong>TrIdent - Transduction Identification</strong> <a href="https://jlmaier12.github.io/TrIdent/"><img src="reference/figures/logo.png" align="right" height="139" alt="TrIdent website"></a></p>
<p><strong>TrIdent automates the analysis of transductomics data by detecting, classifying, and characterizing read coverage patterns associated with potential transduction events.</strong></p>
<p>Transductomics, developed by Kleiner et al. (2020), is a DNA sequencing-based method for the detection and characterization of transduction events in pure cultures and complex communities. Transductomics relies on mapping sequencing reads from a viral-like particle (VLP)-fraction of a sample to contigs assembled from the metagenome (whole-community) of the same sample. Reads from bacterial DNA carried by VLPs will map back to the bacterial contigs of origin creating read coverage patterns indicative of ongoing transduction. <strong>The read coverage patterns detected represent DNA being actively carried or transduced by VLPs. The read coverage patterns do not represent complete transduction events (i.e integration of transduced DNA into new bacterial chromosomes).</strong></p>
<p>To obtain the data needed for transductomics, a microbiome sample of interest is split to prepare two sub-sample types: - Whole-community: Represents the ‘whole-community’ (all bacteria, fungi, virus, etc) in the microbiome of interest - VLP-fraction: Represents only the virus and ‘viral-like particles’ associated with the microbiome of interest - The VLP-fraction must be obtained by an appropriate ultra-purification protocol for your sample type to remove bacterial cells and contaminating free bacterial DNA.</p>
<p>With transductomics and TrIdent, a researcher can obtain information about the phage-host pairs involved in transduction, the types of transduction occuring, and the region of the host genome that is potentially transduced, which allows exploration of transferred genes.</p>
<p><strong>Reference:</strong> Kleiner, M., Bushnell, B., Sanderson, K.E. et al. Transductomics: sequencing-based detection and analysis of transduced DNA in pure cultures and microbial communities. Microbiome 8, 158 (2020). <a href="https://doi.org/10.1186/s40168-020-00935-5" class="external-link uri">https://doi.org/10.1186/s40168-020-00935-5</a></p>
<div class="section level3">
<h3 id="input-files">Input files<a class="anchor" aria-label="anchor" href="#input-files"></a>
</h3>
<p>TrIdent detects read coverage patterns using a pattern-matching algorithm that operates on pileup files. A pileup file is a file format where each row summarizes the ‘pileup’ of reads at specific genomic locations. Pileup files can be used to generate a rolling mean of read coverages and associated base pair positions across a metagenome assembly which reduces data size while preserving read coverage patterns. <strong>TrIdent requires that input pileups files be generated using a 100 bp window/bin size.</strong></p>
<p>Some read mappers, like <a href="https://jgi.doe.gov/data-and-tools/software-tools/bbtools/bb-tools-user-guide/bbmap-guide/" class="external-link">BBMap</a>, will allow for the generation of pileup files in the <a href="https://github.com/BioInfoTools/BBMap/blob/master/sh/bbmap.sh" class="external-link"><code>bbmap.sh</code></a> command with the use of the <code>bincov</code> output with the <code>covbinsize=100</code> parameter/argument. <strong>Otherwise, BBMap’s <a href="https://github.com/BioInfoTools/BBMap/blob/master/sh/pileup.sh" class="external-link"><code>pileup.sh</code></a> can convert .bam files produced by any read mapper to pileup files compatible with TrIdent using the <code>bincov</code> output with <code>binsize=100</code>.</strong></p>
<p>TrIdent requires two pileup files from a transductomics dataset as input:</p>
<ul>
<li>A VLP-fraction pileup: Sequencing reads from a sample’s ultra-purified VLP-fraction mapped to the whole-community metagenome assembly from the same sample.</li>
<li>A whole-community pileup: Sequencing reads from a sample’s whole-community mapped to the whole-community metagenome from the same sample.</li>
</ul>
<p><strong>The data used for each pileup file must originate from the same sample. Pileup files must use a 100 bp window/bin size for the rolling mean.</strong></p>
<p>Transductomics sample preparation, sequencing procedures, and analysis methods are detailed in <a href="https://microbiomejournal.biomedcentral.com/articles/10.1186/s40168-020-00935-5" class="external-link">Kleiner et al. (2020)</a></p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>Install TrIdent with BiocManager:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.github.io/BiocManager/" class="external-link">"BiocManager"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"TrIdent"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jlmaier12/TrIdent" class="external-link">TrIdent</a></span><span class="op">)</span></span></code></pre></div>
<p>Install the development version of TrIdent through Github with BiocManager:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"jlmaier12/TrIdent"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jlmaier12/TrIdent" class="external-link">TrIdent</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="quick-start">Quick Start<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h2>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Load TrIdent</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jlmaier12/TrIdent" class="external-link">TrIdent</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Load sample datasets</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"VLPFractionSamplePileup"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"WholeCommunitySamplePileup"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Run TrIdent:</span></span>
<span><span class="co">## Run first:</span></span>
<span><span class="va">TrIdentOutput</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/TrIdentClassifier.html">TrIdentClassifier</a></span><span class="op">(</span></span>
<span>  VLPpileup <span class="op">=</span> <span class="va">VLPFractionSamplePileup</span>,</span>
<span>  WCpileup <span class="op">=</span> <span class="va">WholeCommunitySamplePileup</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Reformatting pileup files</span></span>
<span><span class="co">#&gt; Starting pattern-matching...</span></span>
<span><span class="co">#&gt; A quarter of the way done with pattern-matching</span></span>
<span><span class="co">#&gt; Half of the way done with pattern-matching</span></span>
<span><span class="co">#&gt; Almost done with pattern-matching!</span></span>
<span><span class="co">#&gt; Determining sizes (bp) of pattern matches</span></span>
<span><span class="co">#&gt; Identifying highly active/abundant or heterogenously integrated</span></span>
<span><span class="co">#&gt;       Prophage-like elements</span></span>
<span><span class="co">#&gt; Finalizing output</span></span>
<span><span class="co">#&gt; Execution time: 14.56secs</span></span>
<span><span class="co">#&gt; 1 contigs were filtered out based on low read coverage</span></span>
<span><span class="co">#&gt; 0 contigs were filtered out based on length</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; HighCovNoPattern        NoPattern    Prophage-like          Sloping </span></span>
<span><span class="co">#&gt;                1                1                4                3</span></span>
<span><span class="co">#&gt; 3 of the prophage-like classifications are highly active or abundant</span></span>
<span><span class="co">#&gt; 1 of the prophage-like classifications are mixed, i.e. heterogenously</span></span>
<span><span class="co">#&gt;         integrated into their bacterial host population</span></span>
<span></span>
<span><span class="co">## Run second:</span></span>
<span><span class="fu"><a href="reference/plotTrIdentResults.html">plotTrIdentResults</a></span><span class="op">(</span></span>
<span>  VLPpileup <span class="op">=</span> <span class="va">VLPFractionSamplePileup</span>,</span>
<span>  WCpileup <span class="op">=</span> <span class="va">WholeCommunitySamplePileup</span>,</span>
<span>  TrIdentResults <span class="op">=</span> <span class="va">TrIdentOutput</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; $NODE_62</span></span></code></pre></div>
<p><img src="reference/figures/README-example-1.png" width="100%"></p>
<div class="sourceCode" id="cb4"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="co">#&gt; $NODE_135</span></span></code></pre></div>
<p><img src="reference/figures/README-example-2.png" width="100%"></p>
<div class="sourceCode" id="cb5"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="co">#&gt; $NODE_1088</span></span></code></pre></div>
<p><img src="reference/figures/README-example-3.png" width="100%"></p>
<div class="sourceCode" id="cb6"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="co">#&gt; $NODE_352</span></span></code></pre></div>
<p><img src="reference/figures/README-example-4.png" width="100%"></p>
<div class="sourceCode" id="cb7"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a><span class="co">#&gt; $NODE_368</span></span></code></pre></div>
<p><img src="reference/figures/README-example-5.png" width="100%"></p>
<div class="sourceCode" id="cb8"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a><span class="co">#&gt; $NODE_560</span></span></code></pre></div>
<p><img src="reference/figures/README-example-6.png" width="100%"></p>
<div class="sourceCode" id="cb9"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a><span class="co">#&gt; $NODE_617</span></span></code></pre></div>
<p><img src="reference/figures/README-example-7.png" width="100%"></p>
<div class="sourceCode" id="cb10"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a><span class="co">#&gt; $NODE_2060</span></span></code></pre></div>
<p><img src="reference/figures/README-example-8.png" width="100%"></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">## Run third:</span></span>
<span><span class="fu"><a href="reference/specializedTransductionID.html">specializedTransductionID</a></span><span class="op">(</span></span>
<span>  VLPpileup <span class="op">=</span> <span class="va">VLPFractionSamplePileup</span>,</span>
<span>  TrIdentResults <span class="op">=</span> <span class="va">TrIdentOutput</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; 2 contigs have potential specialized transduction</span></span>
<span><span class="co">#&gt; We recommend that you also view the results of this search with</span></span>
<span><span class="co">#&gt;       logScale=TRUE</span></span>
<span><span class="co">#&gt; $summaryTable</span></span>
<span><span class="co">#&gt;   contigName specTransduc left right lengthLeft lengthRight</span></span>
<span><span class="co">#&gt; 1    NODE_62          yes  yes    no      45400        &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 2   NODE_135           no   no    no       &lt;NA&gt;        &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 3   NODE_368           no   no    no       &lt;NA&gt;        &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 4   NODE_617          yes  yes   yes      33300        9800</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $Plots</span></span>
<span><span class="co">#&gt; $Plots$NODE_62</span></span></code></pre></div>
<p><img src="reference/figures/README-example-9.png" width="100%"></p>
<div class="sourceCode" id="cb12"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a><span class="co">#&gt; $Plots$NODE_135</span></span></code></pre></div>
<p><img src="reference/figures/README-example-10.png" width="100%"></p>
<div class="sourceCode" id="cb13"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a><span class="co">#&gt; $Plots$NODE_368</span></span></code></pre></div>
<p><img src="reference/figures/README-example-11.png" width="100%"></p>
<div class="sourceCode" id="cb14"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a><span class="co">#&gt; $Plots$NODE_617</span></span></code></pre></div>
<p><img src="reference/figures/README-example-12.png" width="100%"></p>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/jlmaier12/TrIdent/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/jlmaier12/TrIdent/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-2" class="external-link">GPL-2</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing TrIdent</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Jessie Maier <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0009-0001-8575-5386" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Jorden Rabasco <br><small class="roles"> Author, contributor </small> <a href="https://orcid.org/0000-0002-6971-6678" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Craig Gin <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-7447-663X" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Benjamin Callahan <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-8752-117X" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Manuel Kleiner <br><small class="roles"> Author, thesis advisor </small> <a href="https://orcid.org/0000-0001-6904-0287" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jessie Maier, Jorden Rabasco, Craig Gin, Benjamin Callahan, Manuel Kleiner.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
