---
title: "Using TrIdent"
output: rmarkdown::html_vignette
description: >
  TrIdent (Transduction Identification) is a reference-independent tool for the automated detection, classification and characterization of transduction events in   metagenomic sequence data. With only three functions, TrIdent is fast and user-friendly. Learn how to use TrIdent with a sample dataset here. 
vignette: >
  %\VignetteIndexEntry{Using TrIdent}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  fig.width = 9,
  fig.asp = 0.8,
  out.width = "80%",
  collapse = TRUE,
  comment = "#>"
)
```

### Description:
TrIdent automates the transductomics data analysis process. Transductomics is a method developed by Dr. Manuel Kleiner to detect ongoing transduction in DNA sequencing reads. The method relies on mapping reads from a virome (VLP_fraction) of a sample to contigs assembled from the metagenome (whole-community) of the same sample. Reads from bacterial DNA carried by viruses or VLPs (Viral-Like Particles) will map back to their bacterial contigs of origin creating read coverage patterns indicative of active transduction. Read 'Transductomics: sequencing-based detection and analysis of transduced DNA in pure cultures and microbial communities' Kleiner et al. (2020) for more information.

**Reference:**
Kleiner, M., Bushnell, B., Sanderson, K.E. et al. Transductomics: sequencing-based detection and analysis of transduced DNA in pure cultures and microbial communities. Microbiome 8, 158 (2020). https://doi.org/10.1186/s40168-020-00935-5

### Follow this tutorial to learn how TrIdent can be used:

**Input Data:**

The datasets used in this tutorial- 'VLPFraction_sampledata' and 'WholeCommunity_sampledata'- were generated from a conventional mouse fecal metagenome. The homogenized feces represents the whole-community fraction. The VLP-fraction of the fecal sample was separated and purified via CsCl density gradient ultracentrifugation. Both the whole-community and VLP-fraction were shotgun sequenced after which the metagenome was assembled from the whole-community reads. The whole-community and VLP-fraction reads were mapped to the metagenome contigs and two pileup files were generated to summarize the respective read coverages across each contig. A description of this process is detailed in Kleiner et al. (2020). A subset of the 25 largest contigs in the assembly were selected for the sample dataset used in this tutorial. 

The two pileup files were generated using BBMap's `pileup.sh` with a binsize of 100. We **highly** recommend using the commands below to generate the pileup files needed for TrIdent:
```{bash, eval=FALSE}
pileup.sh in=VLPFraction_ReadMappingSorted.bam out=VLPFraction.pileupcovstats bincov=VLPFraction.bincov100 binsize=100 stdev=t

pileup.sh in=WholeCommunity_ReadMappingSorted.bam out=WholeCommunity.pileupcovstats bincov=WholeCommunity.bincov100 binsize=100 stdev=t
```

**In R...**

Load the package into your library:
```{r, eval=FALSE}
library(TrIdent)
```

```{r setup, echo=FALSE, warning=FALSE}
library(TrIdent)
library(knitr)
```

Import your pileup files:

The 'VLPFraction.bincov100' and 'WholeCommunity.bincov100' .txt files generated by BBMap's `pileup.sh` can be used directly for input to TrIdent. The windowsize/binsize used to generate these files **must be 100**. Here is what the raw pileup file should look like, however the information in the first column may be formatted differently depending on your contig accessions:
```{r echo=FALSE, results='asis'}
kable(head(VLPFraction_sampledata), caption="An example of a pileup file generated by BBMap's pileup.sh with binsize=100", format="html")
```

- The first column contains characters strings of the contig accessions given by the assembler used. 
- The second column contains numerical values of read coverages binned over 100bp regions.
- The third column contains integer values for the starting position of each 100bp bin. The position restarts at the start of each new contig.
- The fourth column contains integer values for the starting position of each 100bp bin. The position does NOT restart at the start of each new contig.

TrIdent has built-in data cleaning and reformatting that is guaranteed to work for files output by BBMap's `pileup.sh`. If you do not use BBMap's `pileup.sh` to generate the pileup files, then you are responsible for data-cleaning and reformatting. Your pileup files must be in the following format (same column names, same column classes, etc.) and you must use the `cleanup=FALSE` argument for the `TrIdent_classifier`, `Plot_TrIdentPatternMatches`, and `SpecializedTransduction_ID`.
```{r, echo=FALSE}
kable(head(CleanVLPFraction_sampledata), caption="An example of a cleaned and reformatted pileup file", format="html")
```

###### Running TrIdent:

### Main classifier:
`TrIdent_Classifier` is the main function that TrIdent relies on. This function cleans and reformats your input data, filters contigs based on length and read coverage, performs pattern-matching to classify contigs, identifies active/highly abundant prophage/phage inducible chromosomal islands, and outputs all information in a neat summary table. 

There are four classes that `TrIdent_Classifier` assigns contigs to:

- Gen/Lat/GTA: Generalized, lateral and gene transfer agent (GTA) transduction events. Characterized by sloping read coverage across a contig.
- Prophage-like: Prophage and phage-inducible chromosomal islands(PICIs). Characterized by a 'block' of read coverage on a contig.
- HighVLPWCReadCov: Contigs with no sloping or block patterns, but with high read coverage in the VLP-fraction relative to the read coverage in the whole-community fraction. If the VLP-fraction has a median read coverage at least 
- None: No transduction events.

**Default:**

To run `TrIdent_Classifier` in default-mode, run the following:
```{r}
TrIdent_results <- TrIdent_Classifier(VLPFraction_sampledata, WholeCommunity_sampledata)
```

**Non-default:**

TrIdent resizes the bins or 'windows' used by `pileup.sh` to improve processing time and reduce noise in the data. The default value that TrIdent resizes windows to is 1000bp which we find works the best in most cases. However, depending on the dataset, the user may want to select a different `windowsize`. Users can choose between windowsizes of 200, 500, 1000 or 2000 **only**. We recommend increasing the `windowsize` to 2000 if processing speed is of importance or if data is noisy (i.e. VLP-fraction contaminated with external bacterial DNA). We recommend decreasing the `windowsize` if your data is very clean and you are interested in increasing the resolution of read coverage patterns for the initial classification of contigs. 
 
Increasing/decreasing the `windowsize` may alter the results of the `TrIdent_Classifier` slightly. Note how processing time compares between the default of 1000 `windowsize` and the user-defined 2000 `windowsize`.

Increase `windowsize` to 2000bp:
```{r, eval=FALSE}
TrIdent_results_increasewindow <- TrIdent_Classifier(VLPFraction_sampledata, WholeCommunity_sampledata, windowsize=2000)
```


**Obtain results of main classifier:**

The output of `TrIdent_Classifier` is a list containing five objects:

1. Full_summary_table: A summary table containing the classification information for all contigs that were not filtered out.     
2. Cleaned_summary_table: A cleaned summary table containing the classification information for all contigs classified as either Prophage-like or Gen/Lat/GTA (i.e. 'None classifications removed)
3. PatternMatchInfo: A list of pattern-match info that is used by other functions in TrIdent. 
4. FilteredOutContig_table: A table of contigs that were filtered out and the reason why (either low read coverage or too short(<30kbp)).
5. Windowsize: The windowsize used.

Save the desired list-item to a new variable by using its associated name:
```{r}
summary_table <- TrIdent_results$Full_summary_table
filteredout_contigs <- TrIdent_results$FilteredOutContig_table
```

```{r, echo=FALSE}
kable(head(summary_table), caption="The TrIdent_Classifier output summary table", format="html")
```


```{r, echo=FALSE}
kable(head(filteredout_contigs), caption="The TrIdent_Classifier filtered-out contig summary table", format="html")
```

### Plot results of main classifier
`Plot_TrIdentPatternMatches` will output read coverage plots of all contigs predicted as either Gen/Lat/GTA, Prophage-like, or HighVLPWCReadCov and their respective pattern matches to a list.
```{r}
TrIdent_patternmatches <- Plot_TrIdentPatternMatches(VLPFraction_sampledata, WholeCommunity_sampledata, TrIdent_results)
```

View either all plots at once or one plot at a time. Each plot is named by its respective contig accession.
View all plots:
```{r}
TrIdent_patternmatches
```

View one plot:
```{r}
TrIdent_patternmatches$NODE_1
```



### Identify potential specialized transduction events
`SpecializedTransduction_ID` searches contigs classified as prophage-like for dense read coverage outside the borders of the prophage-like pattern. This function will output results to a summary table 

**Default:**

Search all contigs classified as Prophage-like for specialized transduction:
```{r}
Specialized_transduction <- SpecializedTransduction_ID(VLPFraction_sampledata, TrIdent_results)
```

When you search all contigs, the output of `SpecializedTransduction_ID` will be a list. The first object contains a summary table for the specialized transduction search results:
```{r}
SpecializedTransduction_summary_table <- Specialized_transduction$Summary_table
```
```{r, echo=FALSE}
kable(head(SpecializedTransduction_summary_table), format="html")
```

The second object in the output-list contains a list containing the resulting read coverage plots for all contigs classified as Prophage-like. Each plot is named by the associated contig accession. If `SpecializedTransduction_ID` identifies potential specialized transduction, it will color the plot green whereas if does not identify any specialized transduction, it will color the plot blue.

View all the plots:
```{r}
Specialized_transduction$Plots
```

View a specific plot:
```{r}
Specialized_transduction$Plots$NODE_1
```


**Non-default:**

`SpecializedTransduction_ID` works by starting from the borders of the prophage-like pattern and searching outwards for read coverage that meets the 'requirements' for specialized transduction, as defined by the function. `SpecializedTransduction_ID` first makes sure that any coverage it detects is not disrupted by a defined region of no read coverage (`noreadcov`). The default value for `noreadcov` is 500bp. Secondly, `SpecializedTransduction_ID` ensures that any read coverage it detects outside of a prophage-like shape meets a minimum length requirement (`spectranslength`). The default value for `spectranslength` is 2000bp. By default, `SpecializedTransduction_ID` will search for coverage that is at least 2000bp long and is not interrupted at any point by more than 500bp of no read coverage. 

We suggest using the default values for searching all contigs and only changing the `noreadcov` and `spectranslength` arguments when adapting the specialized transduction search for a specific contig. See example below:

NODE_2 has a small amount of dense read coverage outside of its left border which one might consider potential specialized transduction. `SpecializedTransduction_ID` does not pick this up as potential specialized transduction because there is a region of no read coverage which stops the specialized transduction search. If a user wanted a plot identifying this region as specialized transduction, they might increase the `noreadcov`.
```{r} 
Specialized_transduction_NODE1 <- SpecializedTransduction_ID(VLPFraction_sampledata, TrIdent_results, "NODE_2", noreadcov=900)
Specialized_transduction_NODE1
```


