<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>TrIdent • TrIdent</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="TrIdent">
<meta name="description" content="TrIdent (Transduction Identification) is a reference-independent tool for the  automated detection, classification and characterization of read coverage  patterns in transductomics datasets.
">
<meta property="og:description" content="TrIdent (Transduction Identification) is a reference-independent tool for the  automated detection, classification and characterization of read coverage  patterns in transductomics datasets.
">
<meta property="og:image" content="https://jlmaier12.github.io/TrIdent/logo.png">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">TrIdent</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/TrIdent-vignette.html">TrIdent</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/jlmaier12/TrIdent/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">



<script src="TrIdent-vignette_files/kePrint-0.0.1/kePrint.js"></script><link href="TrIdent-vignette_files/lightable-0.0.1/lightable.css" rel="stylesheet">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>TrIdent</h1>
                        <h4 data-toc-skip class="author">Jessie Maier,
Jorden Rabasco</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/jlmaier12/TrIdent/blob/HEAD/vignettes/TrIdent-vignette.Rmd" class="external-link"><code>vignettes/TrIdent-vignette.Rmd</code></a></small>
      <div class="d-none name"><code>TrIdent-vignette.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p><strong>TrIdent</strong>- <strong>Tr</strong>ansduction
<strong>Ident</strong>ification</p>
<p>TrIdent consists of three main functions which should be run in the
following order:</p>
<ol style="list-style-type: decimal">
<li>
<strong><code><a href="../reference/TrIdentClassifier.html">TrIdentClassifier()</a></code></strong>: Classifies
contigs as ‘Prophage-like’, ‘Sloping’, ‘HighCovNoPattern’, and
‘NoPattern’ using read coverage pattern-matching.</li>
<li>
<strong><code><a href="../reference/plotTrIdentResults.html">plotTrIdentResults()</a></code></strong>: Plots
pattern-matching results of <code><a href="../reference/TrIdentClassifier.html">TrIdentClassifier()</a></code>.</li>
<li>
<strong><code><a href="../reference/specializedTransductionID.html">specializedTransductionID()</a></code></strong>: Searches
contigs classified as Prophage-like by <code><a href="../reference/TrIdentClassifier.html">TrIdentClassifier()</a></code>
for associated specialized transduction events.</li>
</ol>
<p>TrIdent automates the analysis of transductomics data by detecting,
classifying, and characterizing read coverage patterns associated with
potential transduction events. Transductomics, developed by Kleiner et
al.  (2020), is a DNA sequencing-based method for the detection and
characterization of transduction events in pure cultures and complex
communities. Transductomics relies on mapping sequencing reads from a
viral-like particle (VLP)-fraction of a sample to contigs assembled from
the metagenome (whole-community) of the same sample. Reads from
bacterial DNA carried by VLPs will map back to the bacterial contigs of
origin creating read coverage patterns indicative of ongoing
transduction. <strong>The read coverage patterns detected represent DNA
being</strong> <strong>actively carried or transduced by VLPs. The read
coverage patterns do not</strong> <strong>represent complete
transduction events (i.e integration of transduced DNA</strong>
<strong>into new bacterial chromosomes).</strong></p>
<p><strong>Reference:</strong> Kleiner, M., Bushnell, B., Sanderson,
K.E. et al.  Transductomics: sequencing-based detection and analysis of
transduced DNA in pure cultures and microbial communities. Microbiome 8,
158 (2020). <a href="https://doi.org/10.1186/s40168-020-00935-5" class="external-link uri">https://doi.org/10.1186/s40168-020-00935-5</a></p>
<div class="section level3">
<h3 id="acknowledgments">Acknowledgments<a class="anchor" aria-label="anchor" href="#acknowledgments"></a>
</h3>
<p>TrIdent is the result of the combined effort and brain power of many
individuals. Specifically, we would like to thank Dr. Manuel Kleiner,
Dr. Ben Callahan, Dr. Breck Duerkop and Dr. Craig Gin for their
individual expertise and overall support!</p>
</div>
<div class="section level3">
<h3 id="funding">Funding<a class="anchor" aria-label="anchor" href="#funding"></a>
</h3>
<p>The development of TrIdent was supported by a seed grant from the
North Carolina State University Data Science Academy and by the National
Institutes of Health under Award Numbers R35GM138362 and
R01Al171046.</p>
</div>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="section level3">
<h3 id="cran-install">CRAN install<a class="anchor" aria-label="anchor" href="#cran-install"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://pak.r-lib.org/" class="external-link">"pak"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"pak"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu">pak</span><span class="fu">::</span><span class="fu">pak</span><span class="op">(</span><span class="st">"jlmaier12/TrIdent"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jlmaier12/TrIdent" class="external-link">TrIdent</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="github-install">GitHub install<a class="anchor" aria-label="anchor" href="#github-install"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://devtools.r-lib.org/" class="external-link">"devtools"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"jlmaier12/TrIdent"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jlmaier12/TrIdent" class="external-link">TrIdent</a></span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="input-data">Input Data<a class="anchor" aria-label="anchor" href="#input-data"></a>
</h2>
<div class="section level3">
<h3 id="transductomics-data">Transductomics data<a class="anchor" aria-label="anchor" href="#transductomics-data"></a>
</h3>
<p>Transductomics allows for the identification of bacterial DNA being
actively carried or transduced by VLPs. A transductomics dataset
consists of two parts- metagenomes from the whole-community and VLP
fractions of a sample. The whole-community fraction is generated by
extracting and sequencing DNA from the whole sample. The VLP-fraction is
generated by extraction and sequencing DNA of the ultra-purified VLPs in
the sample. VLP ultra-purification is generally done using CsCl
density-gradient ultracentrifugation. Additionally, it is very important
that the VLP-fraction is treated with DNase to remove free DNA! After
sequencing, reads from the whole-community fraction are assembled and
both the whole-community and VLP-fraction reads are mapped to the
assembly. <strong>Read</strong> <strong>mapping should be performed
using a high minimum identity (0.97 or higher)</strong> <strong>and
random mapping of ambiguous reads.</strong> The pileup files needed for
TrIdent are generated using the .bam files produced during read
mapping.</p>
<p><strong>Deep sequencing of the whole-community and VLP-fractions is
needed for</strong> <strong>transductomics! Sample preparation,
sequencing procedures, and</strong> <strong>bioinformatics methods are
detailed in</strong> <strong><a href="https://microbiomejournal.biomedcentral.com/articles/10.1186/s40168-020-00935-5" class="external-link">Kleiner
et al.(2020)</a>.</strong></p>
</div>
<div class="section level3">
<h3 id="pileup-files">Pileup files<a class="anchor" aria-label="anchor" href="#pileup-files"></a>
</h3>
<p>TrIdent detects read coverage patterns using a pattern-matching
algorithm that operates on pileup files. A pileup file is a file format
where each row summarizes the ‘pileup’ of reads at specific genomic
locations. Pileup files can be used to generate a rolling mean of read
coverages and associated base pair positions across a metagenome
assembly which reduces data size while preserving read coverage
patterns. <strong>TrIdent requires that input pileups files be generated
using a 100 bp window/bin size.</strong></p>
<p>Some read mappers, like <a href="https://jgi.doe.gov/data-and-tools/software-tools/bbtools/bb-tools-user-guide/bbmap-guide/" class="external-link">BBMap</a>,
will allow for the generation of pileup files in the <a href="https://github.com/BioInfoTools/BBMap/blob/master/sh/bbmap.sh" class="external-link"><code>bbmap.sh</code></a>
command with the use of the <code>bincov</code> output with the
<code>covbinsize=100</code> parameter/argument. <strong>Otherwise,
BBMap’s</strong> <strong><a href="https://github.com/BioInfoTools/BBMap/blob/master/sh/pileup.sh" class="external-link"><code>pileup.sh</code></a></strong>
<strong>can convert .bam files produced by any read mapper to pileup
files</strong> <strong>compatible with TrIdent using the
<code>bincov</code> output with <code>binsize=100</code>.</strong></p>
<p>TrIdent requires two pileup files from a transductomics dataset as
input:</p>
<ul>
<li>A VLP-fraction pileup: Sequencing reads from a sample’s
ultra-purified VLP-fraction mapped to the sample’s whole-community
metagenome assembly.</li>
<li>A whole-community pileup: Sequencing reads from a sample’s
whole-community mapped to the sample’s whole-community metagenome
assembly.</li>
</ul>
<p><strong>Remember- The data used for each pileup file must originate
from the same</strong> <strong>sample. Pileup files must use a 100 bp
window/bin size for the rolling</strong> <strong>mean.</strong></p>
<p><strong>The input pileup files must have the following
format:</strong></p>
<p>Dataframe with four columns:</p>
<ul>
<li>V1: Contig accession</li>
<li>V2: Mapped read coverage values averaged over 100 bp windows</li>
<li>V3: Starting position (bp) of each 100 bp window. Restarts from 0 at
the start of each new contig.</li>
<li>V4: Starting position (bp) of each 100 bp window. Does NOT restart
at the start of each new contig.</li>
</ul>
<table class="table table" style="margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:left;">
V1
</th>
<th style="text-align:right;">
V2
</th>
<th style="text-align:right;">
V3
</th>
<th style="text-align:right;">
V4
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
NODE_62 length_245921_cov_3.25857_ID_9556347
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
100
</td>
<td style="text-align:right;">
20626396
</td>
</tr>
<tr>
<td style="text-align:left;">
NODE_62 length_245921_cov_3.25857_ID_9556347
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
200
</td>
<td style="text-align:right;">
20626496
</td>
</tr>
<tr>
<td style="text-align:left;">
NODE_62 length_245921_cov_3.25857_ID_9556347
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
300
</td>
<td style="text-align:right;">
20626596
</td>
</tr>
<tr>
<td style="text-align:left;">
NODE_62 length_245921_cov_3.25857_ID_9556347
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
400
</td>
<td style="text-align:right;">
20626696
</td>
</tr>
<tr>
<td style="text-align:left;">
NODE_62 length_245921_cov_3.25857_ID_9556347
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
500
</td>
<td style="text-align:right;">
20626796
</td>
</tr>
<tr>
<td style="text-align:left;">
NODE_62 length_245921_cov_3.25857_ID_9556347
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
600
</td>
<td style="text-align:right;">
20626896
</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section level2">
<h2 id="tridentclassifier">TrIdentClassifier()<a class="anchor" aria-label="anchor" href="#tridentclassifier"></a>
</h2>
<p><code><a href="../reference/TrIdentClassifier.html">TrIdentClassifier()</a></code> is the main function in TrIdent.
This function filters contigs based on length and read coverage,
performs pattern-matching to classify contigs, identifies highly
active/abundant and heterogenously integrated Prophage-like elements,
determines which contigs have high VLP-fraction:whole-community read
coverage ratios, identifies start and stop positions and sizes of
pattern-matches, and calculates slopes for Sloping pattern-matches.</p>
<p>Note that <code><a href="../reference/TrIdentClassifier.html">TrIdentClassifier()</a></code> results should not be
taken at face value and some manual curation is needed! We recommend
careful examination of the output plots to determine if you agree with
the <code><a href="../reference/TrIdentClassifier.html">TrIdentClassifier()</a></code> classifications.</p>
<div class="section level3">
<h3 id="function-components">Function components<a class="anchor" aria-label="anchor" href="#function-components"></a>
</h3>
<div class="section level4">
<h4 id="contig-filtering">Contig filtering<a class="anchor" aria-label="anchor" href="#contig-filtering"></a>
</h4>
<p>Contigs that are too short or have little to no read coverage are
filtered out prior to pattern-matching. <code><a href="../reference/TrIdentClassifier.html">TrIdentClassifier()</a></code>
filters out contigs that do not have at least 10x coverage on a total of
5,000 bp across the whole contig. The read coverage filtering was done
in this way to avoid filtering out long contigs with short Prophage-like
patterns that might get removed if filtering was done with read coverage
averages or medians. Additionally, contigs less than 30,000 bp are
filtered out by default, however this can be changed with the
<code>minContigLength</code> parameter. While contigs shorter than
30,000 bp may be poor quality and are often not long enough to capture
complete transduction patterns, we allow users to set a minimum
<code>minContigLength</code> of 25,000 bp. <strong>If you would like to
reduce the size of your input pileup files for</strong> <strong>TrIdent,
consider pre-filtering your assembly for contigs greater than</strong>
<strong>25,000 bp prior to read mapping!</strong></p>
<p><strong>It is expected that the majority of contigs in the dataset
are filtered out</strong> <strong>due to low read coverage during this
step! If very few contigs are filtered</strong> <strong>out due to low
read coverage, this may be an indication of DNA</strong>
<strong>contamination in your VLP-fraction.</strong></p>
</div>
<div class="section level4">
<h4 id="changing-pileup-windowsize">Changing pileup windowSize<a class="anchor" aria-label="anchor" href="#changing-pileup-windowsize"></a>
</h4>
<p>The input pileup files have 100 bp windows in which the mapped read
coverage is averaged over. This small window size is needed to detect
read coverage patterns associated with specialized transduction which
may only span a few thousand basepairs. However, read coverage patterns
associated with other types of transduction generally don’t require the
resolution that 100 bp windows provide. While users can use the 100 bp
<code>windowSize</code> for <code><a href="../reference/TrIdentClassifier.html">TrIdentClassifier()</a></code>, the
processing time will be increased significantly and noisy data may
interfere with pattern-matching. We find that the default 1,000 bp
<code>windowSize</code> provides a nice balance between processing time
and read coverage pattern resolution.</p>
</div>
<div class="section level4">
<h4 id="pattern-matching">Pattern-matching<a class="anchor" aria-label="anchor" href="#pattern-matching"></a>
</h4>
<p><code><a href="../reference/TrIdentClassifier.html">TrIdentClassifier()</a></code> detects read coverage patterns in
the VLP-fraction using a pattern-matching approach. Several predefined
patterns, described below, are built using the specific length and read
coverage values of the contig being assessed. Patterns are translated
across each contig in 1,000 bp sliding windows and at each translation,
a pattern-match score is calculated by taking the mean absolute
difference of the VLP-fraction read coverage and the pattern values. The
smaller the match-score, the better the pattern-match. After a pattern
is fully translated across a contig, certain aspects of the pattern are
changed (i.e. height, width, slope) and translation is repeated. This
process of translation and pattern re-scaling is repeated until a large
number of pattern variations are tested. After pattern-matching is
complete, the pattern associated with the best match-score is used for
contig classification. Contigs are classified as ‘Prophage-like’,
‘Sloping’, or ‘NoPattern’ during pattern-matching.</p>
<div class="section level5">
<h5 id="patterns">
<strong>Patterns</strong><a class="anchor" aria-label="anchor" href="#patterns"></a>
</h5>
<div class="section level6">
<h6 id="sloping">Sloping:<a class="anchor" aria-label="anchor" href="#sloping"></a>
</h6>
<p>There are four sloping pattern variations in the sloping pattern
class. The sloping patterns are representative of large DNA transfers
that take place during generalized, lateral and gene transfer agent
(GTA) transduction due to the decreasing frequency of DNA packaging
moving away from the packaging initiation sites. During
pattern-matching, the slope values of the sloping patterns are decreased
until a minimum slope of 0.001 (change of 10x read coverage over 10,000
bp) is reached. The minimum slope value can be changed with the
<code>minSlope</code> parameter. Generalized, lateral and GTA
transduction events can span tens to hundreds of kilobasepairs of DNA
and a single contig typically does not capture an entire event.
Depending on which part of the transducing event is captured by the
contig, the sloping can be very severe or almost 0. Patterns 1 and 2
represent contigs that capture a Sloping transducing event somewhere in
the middle of the DNA transfer. Patterns 2 and 4 represent contigs that
capture the jump of read coverage associated with packaging initiation
site of a Sloping transducing event. Patterns 2 and 4 are translated
across the contig in addition to having the slopes changed while only
the slopes are changed on patterns 1 and 2.</p>
<p><img src="TrIdent-vignette_files/figure-html/unnamed-chunk-5-1.png" width="100%"></p>
</div>
<div class="section level6">
<h6 id="prophage-like">Prophage-like:<a class="anchor" aria-label="anchor" href="#prophage-like"></a>
</h6>
<p>There are three block patterns in the Prophage-like pattern class.
The block patterns are representative of integrated genetic elements
that can be excised from the host chromosome and mobilized. The blocks
of read coverage that define this classification are formed when the
reads of mobilized genetic elements packaged by VLPs map back to their
respective integration sites in the host bacterium’s chromosome.
Prophage, phage-inducible chromosomal islands (PICIs), and transposons
are all examples of genetic elements that fall into the Prophage-like
class. During pattern-matching, both the heights and widths of block
patterns are altered and all pattern variations are translated across
the contig. The block pattern widths never get smaller than 10,000 bp by
default, however this can be changed with the <code>minBlockSize</code>
parameter. Pattern 1 represents a Prophage-like element that is entirely
on the contig while patterns 2 and 3 represent Prophage-like elements
that trail off the right or left side of the contig, respectively. While
a Prophage-like classification is not an example of transduction by
itself, there may be transduction associated with Prophage-like
classifications. The improper excision of Prophage-like elements may
lead to the specialized transduction of genes neighboring the element’s
integration site. TrIdent’s <code><a href="../reference/specializedTransductionID.html">specializedTransductionID()</a></code>
function (explained in detail below) will attempt to detect specialized
transduction events associated with Prophage-like classifications.</p>
<p><img src="TrIdent-vignette_files/figure-html/unnamed-chunk-6-1.png" width="100%"></p>
</div>
<div class="section level6">
<h6 id="nopattern">noPattern:<a class="anchor" aria-label="anchor" href="#nopattern"></a>
</h6>
<p>Since the best pattern-match for each contig is determined by
comparing match-scores amongst all pattern-variations from all pattern
classes, we needed a ‘negative control’ pattern to compare against. The
‘NoPattern’ ‘pattern’ serves as a negative control by matching to
contigs with no read coverage patterns. We made two NoPattern patterns
which consist of a horizontal line the same length as the contig being
assessed at either the average or median read coverage for a contig.
This pattern is not re-scaled or translated in any way. Note that read
coverage patterns are heavily dependent on the depth of read coverage
achieved during sequencing and therefore very rare transduction events
may not achieve sufficient read coverage for detection with read
coverage pattern-matching. Rather than label contigs with no read
coverage pattern as having ‘no transduction’, we instead label them as
having ‘no pattern’.</p>
<p><img src="TrIdent-vignette_files/figure-html/unnamed-chunk-7-1.png" width="100%"></p>
</div>
</div>
</div>
<div class="section level4">
<h4 id="highly-activeabundant-and-heterogenously-integratedpresent-prophage-like-elements">Highly active/abundant and heterogenously integrated/present
Prophage-like elements<a class="anchor" aria-label="anchor" href="#highly-activeabundant-and-heterogenously-integratedpresent-prophage-like-elements"></a>
</h4>
<p>Prophage-like elements that are actively replicating or are highly
abundant will typically generate more sequencing reads than the rest of
their host bacterium’s chromosome. This may create a region of elevated
read coverage at the element’s insertion site in the whole-community
fraction read coverage. Conversely, if a Prophage-like element is
integrated into only a portion of the host bacterial population, there
may be a dip or depression in read coverage at the integration site in
the whole-community read coverage. In order to determine if the
whole-community read coverage is elevated or depressed at the site of a
Prophage-like element, one must know the Prophage-like element’s genomic
location. While some tools rely on annotation information to identify
Prophage-like elements in whole-community metagenomes,
<code><a href="../reference/TrIdentClassifier.html">TrIdentClassifier()</a></code> uses the VLP-fraction read coverage
patterns. The locations of Prophage-like pattern-matches are used to
calculate the Prophage-like:non-Prophage-like whole-community read
coverage ratio. Prophage-like patterns with whole-community read
coverage ratios greater than 1.15 are labeled as ‘elevated’ while ratios
less then 0.75 are labeled as ‘depressed’.</p>
</div>
<div class="section level4">
<h4 id="nopattern-classifications-with-high-vlp-fractionwhole-community-read-coverage-ratios">NoPattern classifications with high VLP-fraction:whole-community
read coverage ratios<a class="anchor" aria-label="anchor" href="#nopattern-classifications-with-high-vlp-fractionwhole-community-read-coverage-ratios"></a>
</h4>
<p>If a contig receives a noPattern classification, it proceeds to an
additional classification step which either leaves the classification as
is or re-classifies the contig as having a high
VLP-fraction:whole-community read coverage ratio (‘HighCovNoPattern’).
In other words, contigs with the HighCovNoPattern classification have
even read coverage across the contig (i.e. no notable read coverage
pattern) and the VLP-fraction read coverage is notably higher than the
whole-community read coverage. This re-classification is necessary
because certain biologically significant events that may be of interest
to the user can have these characteristics. First, certain transduction
events, specifically those associated with the Sloping classification,
may be so long that they extend across multiple contigs. The ‘tails’ of
the sloping patterns created by these events may have little to no
actual sloping (no pattern) but relatively high read coverage (Fig S2
Kleiner et al., 2020). Second, there is evidence that membrane vesicle
(MV) mediated transduction (also known as vesiduction or protected
extracellular transformation) produces relatively even read coverage
patterns when purified MV sequencing reads are mapped back to their
bacterial chromosome of origin (Faddetta et al., 2022). Lastly, phage
genomes that have assembled into contigs in the whole-community fraction
may generate high levels of even read coverage as the phage reads in the
VLP-fraction map back to their own genome sequences. Contigs with median
VLP-fraction:Whole-community read coverage ratios greater than 2, in
other words contigs where the median VLP-fraction read coverage value is
2x the whole-community median read coverage value, are re-classified as
HighCovNoPattern.</p>
<p><strong>Reference:</strong> Faddetta, T., Vassallo, A., Del Duca, S.
et al. Unravelling the DNA sequences carried by Streptomyces coelicolor
membrane vesicles. Sci Rep 12, 16651 (2022). <a href="https://doi.org/10.1038/s41598-022-21002-z" class="external-link uri">https://doi.org/10.1038/s41598-022-21002-z</a></p>
</div>
</div>
<div class="section level3">
<h3 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h3>
<p>Default arguments:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">TrIdentOutput</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/TrIdentClassifier.html">TrIdentClassifier</a></span><span class="op">(</span></span>
<span>  VLPpileup <span class="op">=</span> <span class="va">VLPFractionSamplePileup</span>,</span>
<span>  WCpileup <span class="op">=</span> <span class="va">WholeCommunitySamplePileup</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Reformatting pileup files</span></span>
<span><span class="co">#&gt; Starting pattern-matching...</span></span>
<span><span class="co">#&gt; A quarter of the way done with pattern-matching</span></span>
<span><span class="co">#&gt; Half of the way done with pattern-matching</span></span>
<span><span class="co">#&gt; Almost done with pattern-matching!</span></span>
<span><span class="co">#&gt; Determining sizes (bp) of pattern matches</span></span>
<span><span class="co">#&gt; Identifying highly active/abundant or heterogenously integrated</span></span>
<span><span class="co">#&gt;         Prophage-like elements</span></span>
<span><span class="co">#&gt; Finalizing output</span></span>
<span><span class="co">#&gt; Execution time: 23.19secs</span></span>
<span><span class="co">#&gt; 1 contigs were filtered out based on low read coverage</span></span>
<span><span class="co">#&gt; 0 contigs were filtered out based on length</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; HighCovNoPattern        NoPattern    Prophage-like          Sloping </span></span>
<span><span class="co">#&gt;                1                1                4                3</span></span>
<span><span class="co">#&gt; 3 of the prophage-like classifications are highly active or abundant</span></span>
<span><span class="co">#&gt; 1 of the prophage-like classifications are mixed, i.e. heterogenously</span></span>
<span><span class="co">#&gt;         integrated into their bacterial host population</span></span></code></pre></div>
<p><img src="TrIdent-vignette_files/figure-html/unnamed-chunk-8-1.png" width="100%"></p>
</div>
<div class="section level3">
<h3 id="argumentsparameters">Arguments/Parameters<a class="anchor" aria-label="anchor" href="#argumentsparameters"></a>
</h3>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/TrIdentClassifier.html">TrIdentClassifier</a></span><span class="op">(</span><span class="va">VLPpileup</span>, <span class="va">WCpileup</span>,</span>
<span>  windowSize <span class="op">=</span> <span class="fl">1000</span>, minBlockSize <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>  maxBlockSize <span class="op">=</span> <span class="cn">Inf</span>, minContigLength <span class="op">=</span> <span class="fl">30000</span>, minSlope <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>  suggFiltThresh <span class="op">=</span> <span class="cn">FALSE</span>, <span class="va">SaveFilesTo</span></span>
<span><span class="op">)</span></span></code></pre></div>
<ul>
<li>
<code>VLPpileup</code>: VLP-fraction pileup file.</li>
<li>
<code>WCpileup</code>: Whole-community pileup file.</li>
<li>
<code>windowSize</code>: The number of basepairs to average read
coverage values over. Options are 100, 200, 500, 1000 ONLY. Default is
1000.</li>
<li>
<code>minBlockSize</code>: The minimum size (in bp) of the
Prophage-like block pattern. Default is 10000. Must be greater than
1000.</li>
<li>
<code>maxBlockSize</code>: The maximum size (in bp) of the
Prophage-like block pattern. Default is NA (no maximum).</li>
<li>
<code>minContigLength</code>: The minimum contig size (in bp) to
perform pattern-matching on. Must be at least 25000. Default is
30000.</li>
<li>
<code>minSlope</code>: The minimum slope value to test for sloping
patterns. Default is 0.001 (i.e minimum change of 10x read coverage over
100,000 bp).</li>
<li>
<code>suggFiltThresh</code>: TRUE or FALSE, Suggest a filtering
threshold for TrIdent classifications based on the normalized
pattern-match scores. Default is FALSE.</li>
<li>
<code>SaveFilesTo</code>: Optional, Provide a path to the directory
you wish to save output to. A folder will be made within the provided
directory to store results.</li>
</ul>
</div>
<div class="section level3">
<h3 id="output">Output<a class="anchor" aria-label="anchor" href="#output"></a>
</h3>
<p><code><a href="../reference/TrIdentClassifier.html">TrIdentClassifier()</a></code> outputs a histogram displaying the
overall abundance and quality of pattern-matches in addition to the
composition of classifications. The displayed pattern-match scores are
normalized by dividing each score by its associated contig length. The
scores are normalized to visualize the overall quality of
pattern-matching for the entire dataset. Remember, smaller pattern-match
scores correspond to better pattern-matches.</p>
<p>The output of <code><a href="../reference/TrIdentClassifier.html">TrIdentClassifier()</a></code> is a list containing
five objects:</p>
<ol style="list-style-type: decimal">
<li>SummaryTable: A table containing the classification information for
all contigs that were not filtered out.</li>
<li>CleanedSummaryTable: A cleaned summary table containing the
classification information for all contigs classified as Prophage-like,
Sloping and HighCovNoPattern.</li>
<li>PatternMatchInfo: A list of pattern-match information that is used
by other functions in TrIdent.</li>
<li>FilteredOutContigTable: A table containing contigs that were
filtered out and the reason why (low read coverage or too short).</li>
<li>windowSize: The <code>windowSize</code> used.</li>
</ol>
<p>Save the desired list-item to a new variable using its associated
name.</p>
<p>Summary table:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">TrIdentSummaryTable</span> <span class="op">&lt;-</span> <span class="va">TrIdentOutput</span><span class="op">$</span><span class="va">SummaryTable</span></span></code></pre></div>
<table class="table table" style="margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:left;">
contigName
</th>
<th style="text-align:left;">
classifications
</th>
<th style="text-align:right;">
normMatchScore
</th>
<th style="text-align:right;">
VLPWCRatio
</th>
<th style="text-align:right;">
matchSize
</th>
<th style="text-align:right;">
startPosBp
</th>
<th style="text-align:right;">
endPosBp
</th>
<th style="text-align:left;">
proLikeWCReadCov
</th>
<th style="text-align:right;">
proLikeWCReadCovRatio
</th>
<th style="text-align:right;">
slope
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
NODE_62
</td>
<td style="text-align:left;">
Prophage-like
</td>
<td style="text-align:right;">
0.1428571
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
171000
</td>
<td style="text-align:right;">
62000
</td>
<td style="text-align:right;">
233000
</td>
<td style="text-align:left;">
Elevated
</td>
<td style="text-align:right;">
1.4880
</td>
<td style="text-align:right;">
NA
</td>
</tr>
<tr>
<td style="text-align:left;">
NODE_135
</td>
<td style="text-align:left;">
Prophage-like
</td>
<td style="text-align:right;">
0.2737766
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
32000
</td>
<td style="text-align:right;">
149000
</td>
<td style="text-align:right;">
181000
</td>
<td style="text-align:left;">
Elevated
</td>
<td style="text-align:right;">
1.2795
</td>
<td style="text-align:right;">
NA
</td>
</tr>
<tr>
<td style="text-align:left;">
NODE_1088
</td>
<td style="text-align:left;">
Sloping
</td>
<td style="text-align:right;">
0.0802549
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
63000
</td>
<td style="text-align:right;">
1000
</td>
<td style="text-align:right;">
64000
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
0.0024
</td>
</tr>
<tr>
<td style="text-align:left;">
NODE_352
</td>
<td style="text-align:left;">
Sloping
</td>
<td style="text-align:right;">
0.1829770
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
121000
</td>
<td style="text-align:right;">
1000
</td>
<td style="text-align:right;">
122000
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
-0.0001
</td>
</tr>
<tr>
<td style="text-align:left;">
NODE_368
</td>
<td style="text-align:left;">
Prophage-like
</td>
<td style="text-align:right;">
0.1530534
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
30000
</td>
<td style="text-align:right;">
26000
</td>
<td style="text-align:right;">
56000
</td>
<td style="text-align:left;">
Depressed
</td>
<td style="text-align:right;">
0.4001
</td>
<td style="text-align:right;">
NA
</td>
</tr>
<tr>
<td style="text-align:left;">
NODE_560
</td>
<td style="text-align:left;">
HighCovNoPattern
</td>
<td style="text-align:right;">
0.0694395
</td>
<td style="text-align:right;">
16.6016
</td>
<td style="text-align:right;">
95000
</td>
<td style="text-align:right;">
1000
</td>
<td style="text-align:right;">
96000
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
NA
</td>
</tr>
<tr>
<td style="text-align:left;">
NODE_617
</td>
<td style="text-align:left;">
Prophage-like
</td>
<td style="text-align:right;">
0.1613141
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
48000
</td>
<td style="text-align:right;">
34000
</td>
<td style="text-align:right;">
82000
</td>
<td style="text-align:left;">
Elevated
</td>
<td style="text-align:right;">
1.7726
</td>
<td style="text-align:right;">
NA
</td>
</tr>
<tr>
<td style="text-align:left;">
NODE_1401
</td>
<td style="text-align:left;">
NoPattern
</td>
<td style="text-align:right;">
0.1006696
</td>
<td style="text-align:right;">
0.0192
</td>
<td style="text-align:right;">
54000
</td>
<td style="text-align:right;">
1000
</td>
<td style="text-align:right;">
55000
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
NA
</td>
</tr>
<tr>
<td style="text-align:left;">
NODE_2060
</td>
<td style="text-align:left;">
Sloping
</td>
<td style="text-align:right;">
0.1037661
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
27000
</td>
<td style="text-align:right;">
1000
</td>
<td style="text-align:right;">
28000
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
0.0275
</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section level2">
<h2 id="plottridentresults">plotTrIdentResults()<a class="anchor" aria-label="anchor" href="#plottridentresults"></a>
</h2>
<p><code><a href="../reference/plotTrIdentResults.html">plotTrIdentResults()</a></code> allows users to visualize both the
whole-community and VLP-fraction read coverage and the pattern-match
associated with each contig classified as Prophage-like, Sloping and
HighCovNoPattern.</p>
<div class="section level3">
<h3 id="function-components-1">Function components<a class="anchor" aria-label="anchor" href="#function-components-1"></a>
</h3>
<div class="section level4">
<h4 id="re-building-pattern-matches">Re-building pattern-matches<a class="anchor" aria-label="anchor" href="#re-building-pattern-matches"></a>
</h4>
<p>The <code><a href="../reference/TrIdentClassifier.html">TrIdentClassifier()</a></code> output contains information
needed to re-build each pattern-match used for contig classification. To
re-build a complete pattern-match for visualization,
<code><a href="../reference/plotTrIdentResults.html">plotTrIdentResults()</a></code> uses the pattern-match’s minimum and
maximum values and the start and stop positions.</p>
</div>
<div class="section level4">
<h4 id="plotting-read-coverage-and-associated-pattern-matches">Plotting read coverage and associated pattern-matches<a class="anchor" aria-label="anchor" href="#plotting-read-coverage-and-associated-pattern-matches"></a>
</h4>
<p>The whole-community and VLP-fraction read coverage are plotted for
each contig classified as Prophage-like, Sloping or HighCovNoPattern.
The pattern-match associated with the classification is overlaid on the
VLP-fraction read coverage. The whole-community read coverage is
displayed for visualization of elevated or depressed Prophage-like
region read coverage.</p>
</div>
</div>
<div class="section level3">
<h3 id="usage-1">Usage<a class="anchor" aria-label="anchor" href="#usage-1"></a>
</h3>
<p>Default arguments:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">TrIdentPlots</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotTrIdentResults.html">plotTrIdentResults</a></span><span class="op">(</span></span>
<span>  VLPpileup <span class="op">=</span> <span class="va">VLPFractionSamplePileup</span>,</span>
<span>  WCpileup <span class="op">=</span> <span class="va">WholeCommunitySamplePileup</span>,</span>
<span>  TrIdentResults <span class="op">=</span> <span class="va">TrIdentOutput</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="argumentsparameters-1">Arguments/Parameters<a class="anchor" aria-label="anchor" href="#argumentsparameters-1"></a>
</h3>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotTrIdentResults.html">plotTrIdentResults</a></span><span class="op">(</span></span>
<span>  <span class="va">VLPpileup</span>, <span class="va">WCpileup</span>, <span class="va">TrIdentResults</span>, <span class="va">matchScoreFilter</span>,</span>
<span>  <span class="va">saveFilesTo</span></span>
<span><span class="op">)</span></span></code></pre></div>
<ul>
<li>
<code>VLPpileup</code>: VLP-fraction pileup file.</li>
<li>
<code>WCpileup</code>: Whole-community pileup file.</li>
<li>
<code>TrIdentResults</code>: The output from
<code><a href="../reference/TrIdentClassifier.html">TrIdentClassifier()</a></code>
</li>
<li>
<code>matchScoreFilter</code>: Optional, Filter plots using the
normalized pattern match-scores. A suggested filtering threshold is
provided by <code><a href="../reference/TrIdentClassifier.html">TrIdentClassifier()</a></code> if
<code>suggFiltThresh=TRUE</code>.</li>
<li>
<code>saveFilesTo</code>: Optional, Provide a path to the directory
you wish to save output to. A folder will be made within the provided
directory to store results.</li>
</ul>
</div>
<div class="section level3">
<h3 id="output-1">Output<a class="anchor" aria-label="anchor" href="#output-1"></a>
</h3>
<p>The output of <code><a href="../reference/plotTrIdentResults.html">plotTrIdentResults()</a></code> is a list containing
ggplot objects. The list contains all read coverage plots for contigs
classified as Sloping, Prophage-like, or HighCovNoPattern and their
respective pattern-matches.</p>
<p><strong>By default, the plots are displayed with raw read coverage
values. We recommend that users also view plots using
<code>logScale=TRUE</code> as some specialized transduction patterns
occur at such low frequencies they can only be visualized using log
scaled read coverage values.</strong></p>
<p>View all plots:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">TrIdentPlots</span></span></code></pre></div>
<p><img src="TrIdent-vignette_files/figure-html/unnamed-chunk-15-1.png" width="100%"><img src="TrIdent-vignette_files/figure-html/unnamed-chunk-15-2.png" width="100%"><img src="TrIdent-vignette_files/figure-html/unnamed-chunk-15-3.png" width="100%"><img src="TrIdent-vignette_files/figure-html/unnamed-chunk-15-4.png" width="100%"><img src="TrIdent-vignette_files/figure-html/unnamed-chunk-15-5.png" width="100%"><img src="TrIdent-vignette_files/figure-html/unnamed-chunk-15-6.png" width="100%"><img src="TrIdent-vignette_files/figure-html/unnamed-chunk-15-7.png" width="100%"><img src="TrIdent-vignette_files/figure-html/unnamed-chunk-15-8.png" width="100%"></p>
<p>View one plot:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">TrIdentPlots</span><span class="op">$</span><span class="va">NODE_62</span></span></code></pre></div>
<p><img src="TrIdent-vignette_files/figure-html/unnamed-chunk-16-1.png" width="100%"></p>
</div>
</div>
<div class="section level2">
<h2 id="specializedtransductionid">specializedTransductionID()<a class="anchor" aria-label="anchor" href="#specializedtransductionid"></a>
</h2>
<p>Specialized transduction occurs when an integrated Prophage-like
element both excises and packages a small portion of bacterial DNA
directly outside the borders of its insertion site.
<code><a href="../reference/specializedTransductionID.html">specializedTransductionID()</a></code> searches contigs classified as
Prophage-like for dense read coverage outside the borders of the
Prophage-like patterns that might represent specialized transduction
events. Because specialized transduction tends to be fairly short (a
couple kbps) compared to generalized, lateral or GTA transduction (tens
to hundreds of kbps), averaging read coverage over a 1,000 bp distance
(i.e using a <code>windowSize=1000</code>) can ‘blur’ specialized
transduction patterns. This is why specialized transduction is not
identified in <code><a href="../reference/TrIdentClassifier.html">TrIdentClassifier()</a></code>. Instead, we use the
locations of Prophage-like elements identified with
<code><a href="../reference/TrIdentClassifier.html">TrIdentClassifier()</a></code> to guide our search for specialized
transduction in <code><a href="../reference/specializedTransductionID.html">specializedTransductionID()</a></code>.</p>
<p>Note that <code><a href="../reference/specializedTransductionID.html">specializedTransductionID()</a></code> results should not
be taken at face value and some manual curation is needed! We recommend
careful examination of the output plots to determine if you agree with
the <code><a href="../reference/specializedTransductionID.html">specializedTransductionID()</a></code> results. Setting
<code>logScale=TRUE</code> will plot the read coverage in log10 scale
and may help with visualization of low-frequency specialized
transduction events. Be aware that in some cases, log scaled read
coverage can obfuscate the Prophage-like element pattern itself.</p>
<div class="section level3">
<h3 id="function-components-2">Function components<a class="anchor" aria-label="anchor" href="#function-components-2"></a>
</h3>
<div class="section level4">
<h4 id="zoom-in-on-prophage-like-elements">Zoom-in on Prophage-like elements<a class="anchor" aria-label="anchor" href="#zoom-in-on-prophage-like-elements"></a>
</h4>
<p>Since, specialized transduction read coverage patterns are relatively
small, <code><a href="../reference/specializedTransductionID.html">specializedTransductionID()</a></code> ‘zooms-in’ on
Prophage-like elements to aid with specialized transduction
visualization. To ‘zoom-in’ on Prophage-like elements,
<code><a href="../reference/specializedTransductionID.html">specializedTransductionID()</a></code> uses the pattern-match
information generated in <code><a href="../reference/TrIdentClassifier.html">TrIdentClassifier()</a></code> to subset the
contig 50,000 bp outside the pattern-match borders.</p>
</div>
<div class="section level4">
<h4 id="identify-borders-of-prophage-like-elements">Identify borders of Prophage-like elements<a class="anchor" aria-label="anchor" href="#identify-borders-of-prophage-like-elements"></a>
</h4>
<p>Since the <code>windowSize</code> used for pattern-matching in
<code><a href="../reference/TrIdentClassifier.html">TrIdentClassifier()</a></code> is <em>likely</em> larger than 100 bp,
the Prophage-like pattern-match border locations will not directly
translate to the border locations at the 100 bp <code>windowSize</code>.
To re-identify the borders of Prophage-like elements, we use a rolling
standard deviation (SD) to detect large changes in read coverage
associated with Prophage-like element borders. First, the contig is
subset surrounding the left and right pattern-match borders. For contigs
that have a Prophage-like pattern-match that trails off the side of a
contig, then only the border that falls on the contig is subset. Next,
for each subset, the rolling SD is calculated between each read coverage
value and the contig position associated with the largest SD is chosen
as the border location. The border locations are marked on the output
plots with black vertical lines.</p>
</div>
<div class="section level4">
<h4 id="search-for-specialized-transduction-outside-prophage-like-borders">Search for specialized transduction outside Prophage-like
borders<a class="anchor" aria-label="anchor" href="#search-for-specialized-transduction-outside-prophage-like-borders"></a>
</h4>
<p>Once the Prophage-like element borders are identified,
<code><a href="../reference/specializedTransductionID.html">specializedTransductionID()</a></code> starts from the border
locations and searches outwards for dense read coverage that meet the
‘requirements’ for specialized transduction as defined by two parameters
in <code><a href="../reference/specializedTransductionID.html">specializedTransductionID()</a></code>:</p>
<ul>
<li><code>noReadCov</code></li>
<li><code>specTransLength</code></li>
</ul>
<p><code><a href="../reference/specializedTransductionID.html">specializedTransductionID()</a></code> first makes sure that any
coverage it detects outside the borders is not disrupted by a defined
region of no read coverage (<code>noReadCov</code>). The default value
for <code>noReadCov</code> is 500 bp. Secondly,
<code>specializedTransductionID</code> ensures that any read coverage it
detects outside of the Prophage-like borders meets a minimum length
requirement (<code>specTransLength</code>). The default value for
<code>specTransLength</code> is 2000 bp. So by default,
<code><a href="../reference/specializedTransductionID.html">specializedTransductionID()</a></code> will search for coverage
directly outside of the Prophage-like borders that is at least 2000 bp
long and is not interrupted at any point by more than 500 bp of no read
coverage. If these requirements are met,
<code><a href="../reference/specializedTransductionID.html">specializedTransductionID()</a></code> will label the contig as having
specialized transduction. The ‘end’ of the specialized transduction
region is marked on the output plots with red vertical lines.
Additionally, read coverage plots for contigs with specialized
transduction are colored green instead of blue.</p>
</div>
</div>
<div class="section level3">
<h3 id="usage-2">Usage<a class="anchor" aria-label="anchor" href="#usage-2"></a>
</h3>
<p>Default arguments:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">SpecTransduc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/specializedTransductionID.html">specializedTransductionID</a></span><span class="op">(</span></span>
<span>  VLPpileup <span class="op">=</span> <span class="va">VLPFractionSamplePileup</span>,</span>
<span>  TrIdentResults <span class="op">=</span> <span class="va">TrIdentOutput</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; 2 contigs have potential specialized transduction</span></span>
<span><span class="co">#&gt; We recommend that you also view the results of this search with</span></span>
<span><span class="co">#&gt;       logScale=TRUE</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="argumentsparameters-2">Arguments/Parameters<a class="anchor" aria-label="anchor" href="#argumentsparameters-2"></a>
</h3>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/specializedTransductionID.html">specializedTransductionID</a></span><span class="op">(</span><span class="va">VLPpileup</span>, <span class="va">TrIdentResults</span>,</span>
<span>  noReadCov <span class="op">=</span> <span class="fl">500</span>, specTransLength <span class="op">=</span> <span class="fl">2000</span>,</span>
<span>  logScale <span class="op">=</span> <span class="cn">FALSE</span>, <span class="va">matchScoreFilter</span>,</span>
<span>  <span class="va">SaveFilesTo</span>, <span class="va">specificContig</span></span>
<span><span class="op">)</span></span></code></pre></div>
<ul>
<li>
<code>VLPpileup</code>: VLP-fraction pileup file.</li>
<li>
<code>TrIdentResults</code>: The output from
<code><a href="../reference/TrIdentClassifier.html">TrIdentClassifier()</a></code>.</li>
<li>
<code>noReadCov</code>: Number of basepairs of zero read coverage
encountered before specialized transduction searching stops. Default is
500. Must be at least 100.</li>
<li>
<code>specTransLength</code>: Number of basepairs of non-zero read
coverage needed for specialized transduction to be considered. Default
is 2000. Must be at least</li>
</ul>
<ol start="100" style="list-style-type: decimal"><li>
</li></ol>
<ul>
<li>
<code>logScale</code>: TRUE or FALSE, display VLP-fraction read
coverage in log10 scale. Default is FALSE.</li>
<li>
<code>matchScoreFilter</code>: Optional, Filter plots using the
normalized pattern match-scores. A suggested filtering threshold is
provided by <code><a href="../reference/TrIdentClassifier.html">TrIdentClassifier()</a></code> if
<code>suggFiltThresh=TRUE</code>.</li>
<li>
<code>SaveFilesTo</code>: Optional, Provide a path to the directory
you wish to save output to. A folder will be made within the provided
directory to store results.</li>
<li>
<code>specificContig</code>: Optional, Search a specific contig
classified as Prophage-like (“NODE_1”).</li>
</ul>
</div>
<div class="section level3">
<h3 id="output-2">Output<a class="anchor" aria-label="anchor" href="#output-2"></a>
</h3>
<p>The output of <code><a href="../reference/specializedTransductionID.html">specializedTransductionID()</a></code> is a list
containing two objects:</p>
<ol style="list-style-type: decimal">
<li>summaryTable: A table containing the
<code><a href="../reference/specializedTransductionID.html">specializedTransductionID()</a></code> results for all contigs
classified as Prophage-like in <code><a href="../reference/TrIdentClassifier.html">TrIdentClassifier()</a></code>.</li>
<li>Plots: A list of ggplot objects for all contigs classified as
Prophage-like. The plots are ‘zoomed-in’ on the Prophage-like pattern to
aid with specialized transduction visualization. The Prophage-like
element borders identified by <code><a href="../reference/specializedTransductionID.html">specializedTransductionID()</a></code>
are marked on each plot with black vertical lines. If
<code><a href="../reference/specializedTransductionID.html">specializedTransductionID()</a></code> identifies potential
specialized transduction, the plot will be colored green instead of
blue. The end of specialized transduction will be marked with a red
vertical line.</li>
</ol>
<p>Save the desired list-item to a new variable using its associated
name.</p>
<p>Summary table:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">SpecializedTransducSummaryTable</span> <span class="op">&lt;-</span> <span class="va">SpecTransduc</span><span class="op">$</span><span class="va">summaryTable</span></span></code></pre></div>
<table class="table table" style="margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:left;">
contigName
</th>
<th style="text-align:left;">
specTransduc
</th>
<th style="text-align:left;">
left
</th>
<th style="text-align:left;">
right
</th>
<th style="text-align:left;">
lengthLeft
</th>
<th style="text-align:left;">
lengthRight
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
NODE_62
</td>
<td style="text-align:left;">
yes
</td>
<td style="text-align:left;">
yes
</td>
<td style="text-align:left;">
no
</td>
<td style="text-align:left;">
45400
</td>
<td style="text-align:left;">
NA
</td>
</tr>
<tr>
<td style="text-align:left;">
NODE_135
</td>
<td style="text-align:left;">
no
</td>
<td style="text-align:left;">
no
</td>
<td style="text-align:left;">
no
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
</tr>
<tr>
<td style="text-align:left;">
NODE_368
</td>
<td style="text-align:left;">
no
</td>
<td style="text-align:left;">
no
</td>
<td style="text-align:left;">
no
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
</tr>
<tr>
<td style="text-align:left;">
NODE_617
</td>
<td style="text-align:left;">
yes
</td>
<td style="text-align:left;">
yes
</td>
<td style="text-align:left;">
yes
</td>
<td style="text-align:left;">
33300
</td>
<td style="text-align:left;">
9800
</td>
</tr>
</tbody>
</table>
<p>View all plots:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">SpecTransduc</span><span class="op">$</span><span class="va">Plots</span></span></code></pre></div>
<p><img src="TrIdent-vignette_files/figure-html/unnamed-chunk-22-1.png" width="100%"><img src="TrIdent-vignette_files/figure-html/unnamed-chunk-22-2.png" width="100%"><img src="TrIdent-vignette_files/figure-html/unnamed-chunk-22-3.png" width="100%"><img src="TrIdent-vignette_files/figure-html/unnamed-chunk-22-4.png" width="100%"></p>
<p>View a specific plot:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">SpecTransduc</span><span class="op">$</span><span class="va">Plots</span><span class="op">$</span><span class="va">NODE_62</span></span></code></pre></div>
<p><img src="TrIdent-vignette_files/figure-html/unnamed-chunk-23-1.png" width="100%"></p>
</div>
</div>
<div class="section level2">
<h2 id="generate-final-trident-summary-table">Generate final TrIdent summary table<a class="anchor" aria-label="anchor" href="#generate-final-trident-summary-table"></a>
</h2>
<p>If you’d like to combine the summary tables produced by
<code><a href="../reference/TrIdentClassifier.html">TrIdentClassifier()</a></code> and
<code><a href="../reference/specializedTransductionID.html">specializedTransductionID()</a></code>, try the following:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">FinalTrIdentSummaryTable</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html" class="external-link">merge</a></span><span class="op">(</span><span class="va">TrIdentSummaryTable</span>,</span>
<span>  <span class="va">SpecializedTransducSummaryTable</span>,</span>
<span>  by <span class="op">=</span> <span class="st">"contigName"</span>, all.x <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<table class="table table" style="margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:left;">
contigName
</th>
<th style="text-align:left;">
classifications
</th>
<th style="text-align:right;">
normMatchScore
</th>
<th style="text-align:right;">
VLPWCRatio
</th>
<th style="text-align:right;">
matchSize
</th>
<th style="text-align:right;">
startPosBp
</th>
<th style="text-align:right;">
endPosBp
</th>
<th style="text-align:left;">
proLikeWCReadCov
</th>
<th style="text-align:right;">
proLikeWCReadCovRatio
</th>
<th style="text-align:right;">
slope
</th>
<th style="text-align:left;">
specTransduc
</th>
<th style="text-align:left;">
left
</th>
<th style="text-align:left;">
right
</th>
<th style="text-align:left;">
lengthLeft
</th>
<th style="text-align:left;">
lengthRight
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
NODE_1088
</td>
<td style="text-align:left;">
Sloping
</td>
<td style="text-align:right;">
0.0802549
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
63000
</td>
<td style="text-align:right;">
1000
</td>
<td style="text-align:right;">
64000
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
0.0024
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
</tr>
<tr>
<td style="text-align:left;">
NODE_135
</td>
<td style="text-align:left;">
Prophage-like
</td>
<td style="text-align:right;">
0.2737766
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
32000
</td>
<td style="text-align:right;">
149000
</td>
<td style="text-align:right;">
181000
</td>
<td style="text-align:left;">
Elevated
</td>
<td style="text-align:right;">
1.2795
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:left;">
no
</td>
<td style="text-align:left;">
no
</td>
<td style="text-align:left;">
no
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
</tr>
<tr>
<td style="text-align:left;">
NODE_1401
</td>
<td style="text-align:left;">
NoPattern
</td>
<td style="text-align:right;">
0.1006696
</td>
<td style="text-align:right;">
0.0192
</td>
<td style="text-align:right;">
54000
</td>
<td style="text-align:right;">
1000
</td>
<td style="text-align:right;">
55000
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
</tr>
<tr>
<td style="text-align:left;">
NODE_2060
</td>
<td style="text-align:left;">
Sloping
</td>
<td style="text-align:right;">
0.1037661
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
27000
</td>
<td style="text-align:right;">
1000
</td>
<td style="text-align:right;">
28000
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
0.0275
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
</tr>
<tr>
<td style="text-align:left;">
NODE_352
</td>
<td style="text-align:left;">
Sloping
</td>
<td style="text-align:right;">
0.1829770
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
121000
</td>
<td style="text-align:right;">
1000
</td>
<td style="text-align:right;">
122000
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
-0.0001
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
</tr>
<tr>
<td style="text-align:left;">
NODE_368
</td>
<td style="text-align:left;">
Prophage-like
</td>
<td style="text-align:right;">
0.1530534
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
30000
</td>
<td style="text-align:right;">
26000
</td>
<td style="text-align:right;">
56000
</td>
<td style="text-align:left;">
Depressed
</td>
<td style="text-align:right;">
0.4001
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:left;">
no
</td>
<td style="text-align:left;">
no
</td>
<td style="text-align:left;">
no
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
</tr>
<tr>
<td style="text-align:left;">
NODE_560
</td>
<td style="text-align:left;">
HighCovNoPattern
</td>
<td style="text-align:right;">
0.0694395
</td>
<td style="text-align:right;">
16.6016
</td>
<td style="text-align:right;">
95000
</td>
<td style="text-align:right;">
1000
</td>
<td style="text-align:right;">
96000
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
</tr>
<tr>
<td style="text-align:left;">
NODE_617
</td>
<td style="text-align:left;">
Prophage-like
</td>
<td style="text-align:right;">
0.1613141
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
48000
</td>
<td style="text-align:right;">
34000
</td>
<td style="text-align:right;">
82000
</td>
<td style="text-align:left;">
Elevated
</td>
<td style="text-align:right;">
1.7726
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:left;">
yes
</td>
<td style="text-align:left;">
yes
</td>
<td style="text-align:left;">
yes
</td>
<td style="text-align:left;">
33300
</td>
<td style="text-align:left;">
9800
</td>
</tr>
<tr>
<td style="text-align:left;">
NODE_62
</td>
<td style="text-align:left;">
Prophage-like
</td>
<td style="text-align:right;">
0.1428571
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
171000
</td>
<td style="text-align:right;">
62000
</td>
<td style="text-align:right;">
233000
</td>
<td style="text-align:left;">
Elevated
</td>
<td style="text-align:right;">
1.4880
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:left;">
yes
</td>
<td style="text-align:left;">
yes
</td>
<td style="text-align:left;">
no
</td>
<td style="text-align:left;">
45400
</td>
<td style="text-align:left;">
NA
</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="session-information">Session Information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.2.1 (2022-06-23 ucrt)</span></span>
<span><span class="co">#&gt; Platform: x86_64-w64-mingw32/x64 (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Windows 10 x64 (build 19045)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] LC_COLLATE=English_United States.utf8 </span></span>
<span><span class="co">#&gt; [2] LC_CTYPE=English_United States.utf8   </span></span>
<span><span class="co">#&gt; [3] LC_MONETARY=English_United States.utf8</span></span>
<span><span class="co">#&gt; [4] LC_NUMERIC=C                          </span></span>
<span><span class="co">#&gt; [5] LC_TIME=English_United States.utf8    </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] kableExtra_1.4.0   ggplot2_3.5.1      patchwork_1.3.0    knitr_1.49        </span></span>
<span><span class="co">#&gt; [5] TrIdent_0.0.0.9000</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] Rcpp_1.0.13-1      bslib_0.8.0        compiler_4.2.1     pillar_1.9.0      </span></span>
<span><span class="co">#&gt;  [5] jquerylib_0.1.4    tools_4.2.1        digest_0.6.37      viridisLite_0.4.2 </span></span>
<span><span class="co">#&gt;  [9] jsonlite_1.8.9     evaluate_1.0.1     lifecycle_1.0.4    tibble_3.2.1      </span></span>
<span><span class="co">#&gt; [13] gtable_0.3.6       pkgconfig_2.0.3    rlang_1.1.4        cli_3.6.2         </span></span>
<span><span class="co">#&gt; [17] rstudioapi_0.17.1  roll_1.1.7         yaml_2.3.10        pkgdown_2.1.1     </span></span>
<span><span class="co">#&gt; [21] xfun_0.49          fastmap_1.2.0      xml2_1.3.6         withr_3.0.2       </span></span>
<span><span class="co">#&gt; [25] stringr_1.5.1      dplyr_1.1.4        generics_0.1.3     desc_1.4.3        </span></span>
<span><span class="co">#&gt; [29] fs_1.6.5           vctrs_0.6.5        htmlwidgets_1.6.4  sass_0.4.9        </span></span>
<span><span class="co">#&gt; [33] systemfonts_1.1.0  grid_4.2.1         tidyselect_1.2.1   svglite_2.1.3     </span></span>
<span><span class="co">#&gt; [37] glue_1.8.0         R6_2.5.1           textshaping_0.4.1  fansi_1.0.6       </span></span>
<span><span class="co">#&gt; [41] rmarkdown_2.29     purrr_1.0.2        tidyr_1.3.1        farver_2.1.2      </span></span>
<span><span class="co">#&gt; [45] magrittr_2.0.3     scales_1.3.0       htmltools_0.5.8.1  colorspace_2.1-1  </span></span>
<span><span class="co">#&gt; [49] labeling_0.4.3     ragg_1.3.3         utf8_1.2.4         stringi_1.8.4     </span></span>
<span><span class="co">#&gt; [53] RcppParallel_5.1.9 munsell_0.5.1      cachem_1.1.0</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jessie Maier, Jorden Rabasco.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
